src/MongoDB.EntityFrameworkCore/Storage/MongoClientWrapper.cs

```diff
@@ -99,138 +101,278 @@ public async Task<IClientSessionHandle> StartSessionAsync(CancellationToken canc
+                    "Index creation timed out. Please create indexes using MongoDB Compass or the mongosh shell.");
+            }
+        } while (!isReady);
+    }
+
+    private void BuildIndexes(
+        IModel model,
         IEntityType entityType,
-        IMongoIndexManager<BsonDocument> indexManager,
-        List<string> existingIndexNames,
         string[] path,
-        CancellationToken cancellationToken)
+        string collectionName,
+        Dictionary<string, List<string>?> existingIndexesMap,
+        Dictionary<string, List<string>?> existingSearchIndexesMap,
+        Dictionary<string, List<CreateIndexModel<BsonDocument>>?> indexModelsMap,
+        Dictionary<string, List<CreateSearchIndexModel>?> searchIndexModelsMap)
     {
-        foreach (var index in entityType.GetIndexes())
+        var existingIndexes = existingIndexesMap[collectionName]!;
+        var existingSearchIndexes = existingSearchIndexesMap[collectionName]!;
+
+        if (!indexModelsMap.TryGetValue(collectionName, out var indexModels))
+        {
+            indexModels = [];
+            indexModelsMap[collectionName] = indexModels;
+        }
+
+        if (!searchIndexModelsMap.TryGetValue(collectionName, out var searchIndexModels))
+        {
+            searchIndexModels = [];
+            searchIndexModelsMap[collectionName] = searchIndexModels;
+        }
+
+        foreach (var index in entityType.GetIndexes().Where(i => !i.GetVectorIndexOptions().HasValue))
         {
             var name = index.Name ?? MakeIndexName(index, path);
-            if (!existingIndexNames.Contains(name))
+            if (!existingIndexes.Contains(name))
             {
-                await indexManager.CreateOneAsync(CreateIndexDocument(index, name, path), null, cancellationToken)
-                    .ConfigureAwait(false);
+                existingIndexes.Add(name);
+                indexModels!.Add(CreateIndexDocument(index, name, path));
             }
         }
 
         foreach (var key in entityType.GetKeys().Where(k => !k.IsPrimaryKey()))
         {
             var name = MakeIndexName(key, path);
-            if (!existingIndexNames.Contains(name))
+            if (!existingIndexes.Contains(name))
             {
-                await indexManager.CreateOneAsync(CreateKeyIndexDocument(key, name, path), null, cancellationToken)
-                    .ConfigureAwait(false);
+                existingIndexes.Add(name);
+                indexModels!.Add(CreateKeyIndexDocument(key, name, path));
             }
         }
 
-        var ownedEntityTypes = entityType.Model.GetEntityTypes()
-            .Where(o => o.FindDeclaredOwnership()?.PrincipalEntityType == entityType);
+        foreach (var index in entityType.GetIndexes().Where(i => i.GetVectorIndexOptions().HasValue))
+        {
+            var options = index.GetVectorIndexOptions()!.Value;
+            var name = index.Name ?? index.Name ?? index.Properties.Single().Name + "VectorIndex";
```

Comment:
Was reflecting on things last night, and I'm going to change some of this stuff so that the EF model contains the name we generate, when we generate a name. This way we can better choose the correct index in queries later. It also aligns with what other providers do.